### PR TITLE
Add encoder prefetch for heterogeneous MIMO training

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## What changed

- add opt-in, configurable-depth prefetch for frozen MIMO encoder features with any compatible batch iterator.
- keep completed features in a bounded FIFO and cap unfinished encoder GPU work at one batch.
- overlap CPU generation of the next batch with the current GPU encode using one staged CPU batch.
- retain encoder-rank `input_ids` on CPU, removing an unnecessary H2D copy and pre-bridge GPU-to-host synchronization.
- keep the trainable projection on the schedule stream and inside autograd.

## Validation

- final cleanup bundle: Black, isort, Ruff, and 76 focused/surrounding MIMO tests passed on CW DFW `14223654`
- final 8-GPU heterogeneous mock smoke completed two iterations with finite loss/gradients and zero skipped/NaN iterations on CW DFW `14223688`; all eight rank logs were checked

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [ce3994f](https://github.com/NVIDIA/Megatron-LM/pull/5833/commits/ce3994f70c93426631c3508274dcf702ffa9a85c)

File: examples/mimo/pretrain_mimo.py [hunk 1]
```diff
@@ -68,43 +78,71 @@ def main() -> None:
...
+    communicator = provider.build_communicator(args, topology)
 
-        loaders = build_train_valid_test_data_loaders(args, topology)
-        iterators = tuple(iter(loader) if loader is not None else None for loader in loaders)
+    if args.mimo_encoder_prefetch and len(provider.encoder_module_names) != 1:
+        raise ValueError("encoder prefetch requires exactly one encoder")
+
+    captured_model = {}
+    hooks = []
+    if args.mimo_encoder_prefetch:
```
Comment (deepakn94) [inline examples/mimo/pretrain_mimo.py:99]: Can you add a comment here on how this works?

**Review by deepakn94 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [704be29](https://github.com/NVIDIA/Megatron-LM/pull/5833/commits/704be2995c19ac18841de4177593a11852d1aa4a)

**Comment by yashaswikarnati** [comment]
/ok to test [704be29](https://github.com/NVIDIA/Megatron-LM/pull/5833/commits/704be2995c19ac18841de4177593a11852d1aa4a)

**Comment by yashaswikarnati** [comment]
/ok to test [2019209](https://github.com/NVIDIA/Megatron-LM/pull/5833/commits/2019209bf527186f0eb9485caddf018e2a245e4e)
